gspread/utils.py

```diff
@@ -979,6 +985,130 @@ def to_records(
     return [dict(zip(headers, row)) for row in values]
 
 
+def _expand_right(values: List[List[str]], start: int, end: int, row: int) -> int:
+    """This is a private function, returning the column index of the first empty cell
+    on the given row.
+
+    Search starts from ``start`` index column.
+    Search ends on ``end`` index column.
+    Searches only in the row pointed by ``row``.
+
+    If no empty value is found, it will return the given ``end`` index.
+    """
+    for column in range(start, end):
+        # in case the given row is smaller that what is being asked
+        if column >= len(values[row]):
+            return len(values[row]) - 1
+
+        if values[row][column] == "":
+            return column
+
+    return end
+
+
+def _expand_bottom(values: List[List[str]], start: int, end: int, col: int) -> int:
+    """This is a private function, returning the row index of the first empty cell
+    on the given column.
+
+    Search starts from ``start`` index row.
+    Search ends on ``end`` index row.
+    Searches only in the column pointed by ``col``.
+
+    If no empty value is found, it will return the given ``end`` index.
+    """
+    for rows in range(start, end):
+        # in case we try to look further than last row
+        if rows >= len(values):
+            return len(values) - 1
+
+        # this row is smaller than the others, just keep looking
+        if col >= len(values[rows]):
+            continue
+
+        if values[rows][col] == "":
+            return rows
```

Comment:
here I noticed we have rows smaller than the starting row sometimes.
I assumed we should keep looking for rows bellow, but I believe this is wrong.

we should look down for any empty value `""` OR any row smaller than starting row right ?

Reply:
I prefer to answer questions by writing tests. not sure exactly what you mean

Reply:
what I mean is currently we can have a response from the API that looks like this:

||A |B|C|
|-|-| - | -|
|1|v1|v2|v3|
|2|v4|v5| |
|3|v6|v7|v8|

and currently in the code we:
1. start looking right from A1, until C1 (latest value in the row)
2. then we start to look down.
3. when we check C2, the row 2 returned by the API is smaller than the starting row 1
4. currently we just ignore it and keep looking to the row below
5. I believe that's wrong, we should stop. 

So the code `if col >= len(values[rows]):` is wrong we should not continue but instead consider it as an empty cell.

Reply:
yes. With my suggestion (look for full top row & left column), this should not matter(?)